README.md
# miot
Modular Internet of Things - Core modules

STATUS:  PRE-ALPHA  ** UNSTABLE DO NOT INSTALL **

VERSION: 1.0.3.8
    - WARNING - Unstable development version
    - Not feature complete
    - Under development
    - Not ready for internal testing

A Pre-Alpha release:
* Is still under development and considered unstable
* May have had limited or no internal testing.
* May not be feature complete
* May have features that are partially complete or not working. 
* May have known or unknown bugs
* May operate in an unexpected way. 
* May contain output debugging data to logs or screen.
* Is NOT supported: Use at your own risk.

Please do not raise tickets for bugs found in this release.

# INFORMATION

These are the core modules that are used by MIoT.

More information about MIoT is available from it's [project page on github][https://github.com/automation-itspeedway-net/miot].

# FEATURES:

* Central modules for mqtt, logs and config used by other packages
* Adds functionality by means of REPL to 'miot' shell command

# PLATFORMS

    - Linix (Mint)  -   tbc
    - Linux (Other) -   Untested
    - Raspbarian    -   tbc
    - Windows       -   Currently Unsupported
    
# PREREQUISITES

    * [miot][https://github.com/automation-itspeedway-net/miot]
    * python 3
    * pip
    * git

# INSTALLATION

    pip3 install --upgrade miot
    
Check that is it working:

    miot version

If this fails; you probably need to add  ~/.local/bin to your path. This is a common problem on Debian/Ubuntu/Mint because it is not included by default. Run these commands and then try testing it works again.

    echo export PATH="$HOME/.local/bin:$PATH">>~/.bashrc
    source ~/.bashrc

Now that it is working, run setup

    miot setup
    
# UPGRADE

    miot disable
    pip3 install --upgrade miot
    miot enable
    
# KNOWN BUGS

* Pre-Alpha versions will contain known and unknown bugs.


CHANGES.md
# CHANGE CONTROL

DATE         VERSION  STATUS
15 MAR 2019  0.0      Initial Outline
25 MAR 2019  0.1      PLANNING
27 MAR 2019  0.2      PRE-ALPHA

## PLANNING RELEASES

DATE         VERSION  STATUS
25 MAR 2019  0.1.1
    - Revised event handlers
26 MAR 2019  0.1.2
    - Bug fixes / Removed some debug output
26 MAR 2019  0.1.3
    - Added config.getfloat(), config.set() and config.write()

## PRE-ALPHA RELEASES

DATE         VERSION
27 MAR 2019  1.0.0  ** Introduced invalid version number **
    - Added miot-repl
04 JUL 2020  1.0.1
11 JUL 2020  1.0.2
11 JUL 2020  1.0.3
12 JUL 2020  1.0.3.1
    - "miotrepl" included in package
    - Documentation typo's
    - Bug fixes
15 JUL 2020  1.0.3.2
    - Re-write miot.mqtt and miot.log
    - Bug fixes
15 JUL 2020  1.0.3.3
    - Removed debugging statements from miot.config
    - Fixed miotrepl
16 JUL 2020  1.0.3.4
    - Bug fixes
18 JUL 2020  1.0.3.5
    - Bug fixes
25 JUL 2020  1.0.3.6
    - miot.config: Added default mqtt.root variable
    - miot.mqtt: Added lwt()
    - miotrepl:
        Tidied up header
        Added ANSI color to MQTT messages and Errors
        Bug fixes
        Several improvements
26 JUL 2020  1.0.3.7
    - Moved scripts from bin to scripts
    - included miot (shell script)

## ALPHA RELEASES


test/test-log.py
# Test the log module
# (c) Copyright Si Dunford, Jul 2020
#
# VERSION 1.1

from miot import log

# OPTIONAL : Set the log file format
#log.format()
#log.format( log.SYSLOG )
log.format( log.DAEMON )

log.debug( "Test DEBUG Message" )
log.info( "Test INFO Message" )
log.notice( "Test NOTICE Message" )
log.warning( "Test WARNING Message" )
log.error( "Test ERROR Message" )
log.critical( "Test CRITICAL Message" )
log.alert( "Test ALERT Message" )
log.emergency( "Test EMERGENCY Message" )

# Force an exception
try:
    x = 1 / 0
except ZeroDivisionError as e:
    log.exception( 'ZeroDivisionError: %s', e )





test/test-mqtt.py
# Test the MQTT module
# (c) Copyright Si Dunford, Jul 2020
#
# VERSION 1.1

import sys,time
from miot import mqtt

def on_error( broker, error ):
    try:
        err=str(error['id'])
        print( '# ERROR ('+err+'): '+error['message']+", "+broker.error( err ) )
    except Exception as e:
        print(e)

def on_connect( broker, msg ):
    print( "# Connect", str(msg ) )
    print( "- Broker=", broker.hostname )
    
    # Subscribe to topic
    result,mid = broker.subscribe( '#' )
    print( "- Subscribed with mid="+str(mid) )
    
    # Send message
    broker.publish( "miot/event", "test-mqtt.py connected" )

def on_disconnect( broker, msg ):
    print( "# Disconnect", broker.hostname, str(msg ) )

def on_publish( broker, msg ):
    print( "# Publish", broker.hostname, str(msg ) )

def on_subscribe( broker, msg ):
    print( "# Subscribe", broker.hostname, str(msg ) )

def on_unsubscribe( broker, msg ):
    print( "# Unsubscribe", broker.hostname, str(msg ) )

def on_message( broker, msg ):
    print( "# Message", broker.hostname, str(msg ) )
    pass

def main():
    #try:
    print( "* Create broker" )
    broker = mqtt.broker()
    #except Exception as e:
    #    print( e )
    #    sys.exit()

    print( "* Setup authentication" )
    broker.authenticate( 'RunForTheHills', 'AnotherFineMess' )

    # Add event handlers
    print( "* Setup event handlers" )
    broker.on( 'error', on_error )
    broker.on( 'connect', on_connect )
    broker.on( 'disconnect', on_disconnect )
    broker.on( 'publish', on_publish )
    broker.on( 'subscribe', on_subscribe )
    broker.on( 'unsubscribe', on_unsubscribe )
    broker.on( 'message', on_message )

    # Connect to broker
    print( "* Connecting to broker" )
    broker.connect( '192.168.1.228' )

    # AYNC Listener
    #broker.start()     # ASYNC
    # while running:
    #   do something
    # broker.stop()
    
    # SYNC Blocking
    broker.wait()       # SYNC (Blocking)

    # MANUAL Dispatch
    # while True:.
    #   broker.dispatch()
    
    #while True:
    # THIS IS BAD: SEE HERE
    # https://blog.miguelgrinberg.com/post/how-to-make-python-wait
    #    time.sleep(10)

if __name__=='__main__':
    try:
        main()
    except KeyboardInterrupt:
        sys.exit()



test/test-config.py
# Test the config module
# (c) Copyright Si Dunford, Jul 2020
#
# VERSION 1.1

from miot import config

print( config.tostring() )

# TEST CONFIG
print( "RESULT="+config.get( "mqtt.host", "xyz" ) )
print( "RESULT="+config.get( "mqtt.default.host", "abc" ) )
print( "RESULT="+config.get( "rss.feed", "123" ) )

print( config.tostring() )

config.wipe()
config.default()

config.set( "mqtt.host", "192.168.1.228" )
config.set( "mqtt.default.host", "192.168.1.253" )
config.set( "rss.itspeedway", {'url':"https://itspeedway.net"} )
config.set( "rss.bbc", {'url':"https://bbc.co.uk"} )

config.save()

print( config.tostring() )




scripts/miotrepl.py
#!/usr/bin/env python3
#!/usr/bin/python3

# Modular Internet of Things
# Command line utility / REPL
# (c) Copyright Si Dunford, Mar 2019
#
# STATE: Pre-Alpha

import sys
from miot import config,mqtt

VERSION = '1.0.3.6'

brokers = {}
interactive = False

## SUPPORT FUNCTIONS

def die( message, subtext='', pointer=0 ):
    print( COLOR(YELLOW)+message )
    if subtext!='': print( "  "+subtext )
    if pointer>0:
        print( " "*(pointer+1)+"^" )
    if interactive: return
    sys.exit(1)

def repl():
    #print( "Welcome to Modular IoT" )
    print( "REPL: "+str(VERSION) )
    print( "" )
    print( "Type 'quit' to exit, or 'help' for assistance." )
    print( "" )
    running = True
    try:
        while running:
            value = input( COLOR(WHITE)+":" )
            if value!='':
                cmd = value.split(" ")
                if cmd[0]=='quit': 
                    running = False
                else:
                    func = CMDswitcher.get( cmd[0], unknown )
                    func( cmd )
    except KeyboardInterrupt:
        print( COLOR(WHITE)+"" )
        print( "- Bye!" )
        sys.exit()
    except Exception as e:
        print( COLOR(YELLOW)+e )

## ANSI COLOUR SUPPORT

BLACK=0
RED=1
GREEN=2
YELLOW=3
BLUE=4
PURPLE=5
CYAN=6
WHITE=7

def COLOR( fg ):
    return "\033[{}m".format( fg+30 )

## MQTT SUPPORT

def on_error( broker, error ):
    try:
        #print( "Error" )
        err = str(error['id'])
        print( COLOR(YELLOW)+"!ERROR: {} | ({}) {} | {}".format( broker.hostname, err, error['message'], broker.error( err ) ))
    except Exception as e:
        print( COLOR(YELLOW)+e )
        
def on_connect( broker, msg ):
    #print("Connected")
    print( COLOR(GREEN)+"!CONNECTED: {}".format( broker.hostname ) )

def on_message( broker, msg ):
    #print( "Message" )
    print( COLOR(CYAN)+"!MESSAGE: {} | {} | {}".format( broker.hostname, msg['topic'], msg['message'] ))

def connect( host='default' ):
    try:
        global brokers
        brokerkey = host
        if brokerkey=='': brokerkey='default'
        if brokerkey in brokers:
            #print( "- Using saved host" )
            return brokers[brokerkey]
        #
        if( host=='' ):
            hostip   = config.get( "mqtt.host" )
            port     = config.get( "mqtt.port" )
            username = config.get( "mqtt.username" )
            userpass = config.get( "mqtt.password" )
        else:
            hostip   = config.get( "mqtt.{}.host".format( host ) )
            port     = config.get( "mqtt.{}.port".format( host ) )
            username = config.get( "mqtt.{}.username".format( host ) )
            userpass = config.get( "mqtt.{}.password".format( host ) )
        #
        #print( "  HOST:     "+host )
        #print( "  BROKER:   {}:{}".format(hostip,port) )
        #print( "  USERNAME: "+username )
        
        broker = mqtt.broker()
        broker.authenticate( username, userpass )
        broker.on( 'error',   on_error )
        broker.on( 'connect', on_connect )
        broker.on( 'message', on_message )
        broker.connect( hostip, port )
        #
        broker.start()  # Start ASYNC MODE
        #
        brokers[brokerkey]=broker
        return broker
    except Exception as e:
        print( COLOR(YELLOW)+e )

def disconnect( brokerkey ):
    global brokers
    if brokerkey=='': brokerkey='default'
    if brokerkey in brokers:
        brokers[brokerkey].stop()
        del brokers[brokerkey]

def disconnectall():
    global brokers
    for key,value in list(brokers.items()):
        disconnect( key )
        
## COMMANDS

"""
def exist( args ):
    if len(args)<1:
        die( COLOR(YELLOW)+"Missing parameter:", "exist <key>", 7 )
    key = args[0]
    default = False
    print( len(args))
    if len(args)==2:
        print( "has default flag" )
        temp = args[2][0]
        print(temp)
        default = True if args[2][0].lower()==true else False
    result = config.exist( key, default )
    print( COLOR(WHITE)+str(result) )
"""

def get( args ):
    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', cmd+' <key>', len(cmd)+2)
        return
    key = args.pop(0)
    if len(args)>0:
        unexpected = " ".join(args)
        die( COLOR(YELLOW)+'Unexpected argument:', cmd+' '+key+' '+unexpected, len(cmd)+3+len(key)  )
        return
    print( COLOR(WHITE)+key+" = '"+str(config.get( key ))+"'" )
        
def help( args ):
    print( COLOR(WHITE)+"""
  get <key>              Displays a configuration key
  set <key> <value>      Sets a configuration key
  subscribe <topic>      Subscribes to a topic (Shows messages)
  publish <topic> <msg>  Publishes a message on a topic
  version                Shows the current version
  
  Alias:
  show <key>    Same as "get <key>"
  listen <topic>  Same as "subscribe <topic>"
  """)

"""
def list( args ):
    print( COLOR(WHITE)+"LIST" )

    # READ CONFIG.INI FOR INSTALLED COMPONENTS
    # READ MODULES.DAT FOR AVAILABLE COMPONENTS
    # SHOW LIST
"""

def publish( args ):

    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'publish <topic> <Message>', 9 )
        return
        
    topic = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'publish '+topic+' <Message>', 10+len(topic) )
        return
       
    message  = " ".join(args)
        
    broker = connect()
    broker.publish( topic, message )
    print( COLOR(GREEN)+"- Published to {}.".format( topic ) )
 
def set( args ):
    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'set <key> <value>', 5)
        return
    key = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'set '+key+' <value>', 6+len(key))
        return
    value = " ".join(args)
    config.set( key, value )
    config.save()
    print( COLOR(WHITE)+key+" = '"+str(value)+"'" )

def subscribe( args ):
    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', cmd+' <topic>', len(cmd)+2)
        return
    topic = args.pop(0)
    if len(args)>0:
        unexpected = " ".join(args)
        die( COLOR(YELLOW)+'Unexpected argument:', cmd+' '+topic+' '+unexpected, len(cmd)+3+len(topic)  )
        return
        
    broker = connect()
    broker.subscribe( topic )
   
#def unsubscribe( args ):
#    pass

def unknown( args ):
    die( COLOR(YELLOW)+"- Unknown command: {}".format(args[0]) )
     
#def update( args ):
#    print( COLOR(WHITE)+"UPDATE")

def version( args ):
    print( COLOR(WHITE)+"Version: "+VERSION )

# COMMAND SWITCHER
CMDswitcher = { 
    "--help":help,
    "get":get,
    "help":help,
    "listen":subscribe,
    "publish":publish,
    "set":set,
    "show":get,
    "subscribe":subscribe,
    "version":version 
}
# "list":list,
# "update":update, 

if __name__=="__main__":

    # PROCESS ARGUMENTS
    if len(sys.argv) <= 1:
        interactive = True
        repl()
        sys.exit(0)

    func = CMDswitcher.get( sys.argv[1], unknown )
    func( sys.argv[1:] )






scripts/miot.py
#!/usr/bin/env python3
#!/usr/bin/python3

# Modular Internet of Things
# Command line utility / REPL
# (c) Copyright Si Dunford, Mar 2019
#
# STATE: Pre-Alpha

import sys
from miot import config,mqtt

VERSION = '1.0.3.6'

brokers = {}
interactive = False

## SUPPORT FUNCTIONS

def die( message, subtext='', pointer=0 ):
    print( COLOR(YELLOW)+message )
    if subtext!='': print( "  "+subtext )
    if pointer>0:
        print( " "*(pointer+1)+"^" )
    if interactive: return
    sys.exit(1)

def repl():
    #print( "Welcome to Modular IoT" )
    print( "REPL: "+str(VERSION) )
    print( "" )
    print( "Type 'quit' to exit, or 'help' for assistance." )
    print( "" )
    running = True
    try:
        while running:
            value = input( COLOR(WHITE)+":" )
            if value!='':
                cmd = value.split(" ")
                if cmd[0]=='quit': 
                    running = False
                else:
                    func = CMDswitcher.get( cmd[0], unknown )
                    func( cmd )
    except KeyboardInterrupt:
        print( COLOR(WHITE)+"" )
        print( "- Bye!" )
        sys.exit()
    except Exception as e:
        print( COLOR(YELLOW)+e )

## ANSI COLOUR SUPPORT

BLACK=0
RED=1
GREEN=2
YELLOW=3
BLUE=4
PURPLE=5
CYAN=6
WHITE=7

def COLOR( fg ):
    return "\033[{}m".format( fg+30 )

## MQTT SUPPORT

def on_error( broker, error ):
    try:
        #print( "Error" )
        err = str(error['id'])
        print( COLOR(YELLOW)+"!ERROR: {} | ({}) {} | {}".format( broker.hostname, err, error['message'], broker.error( err ) ))
    except Exception as e:
        print( COLOR(YELLOW)+e )
        
def on_connect( broker, msg ):
    #print("Connected")
    print( COLOR(GREEN)+"!CONNECTED: {}".format( broker.hostname ) )

def on_message( broker, msg ):
    #print( "Message" )
    print( COLOR(CYAN)+"!MESSAGE: {} | {} | {}".format( broker.hostname, msg['topic'], msg['message'] ))

def connect( host='default' ):
    try:
        global brokers
        brokerkey = host
        if brokerkey=='': brokerkey='default'
        if brokerkey in brokers:
            #print( "- Using saved host" )
            return brokers[brokerkey]
        #
        if( host=='' ):
            hostip   = config.get( "mqtt.host" )
            port     = config.get( "mqtt.port" )
            username = config.get( "mqtt.username" )
            userpass = config.get( "mqtt.password" )
        else:
            hostip   = config.get( "mqtt.{}.host".format( host ) )
            port     = config.get( "mqtt.{}.port".format( host ) )
            username = config.get( "mqtt.{}.username".format( host ) )
            userpass = config.get( "mqtt.{}.password".format( host ) )
        #
        #print( "  HOST:     "+host )
        #print( "  BROKER:   {}:{}".format(hostip,port) )
        #print( "  USERNAME: "+username )
        
        broker = mqtt.broker()
        broker.authenticate( username, userpass )
        broker.on( 'error',   on_error )
        broker.on( 'connect', on_connect )
        broker.on( 'message', on_message )
        broker.connect( hostip, port )
        #
        broker.start()  # Start ASYNC MODE
        #
        brokers[brokerkey]=broker
        return broker
    except Exception as e:
        print( COLOR(YELLOW)+e )

def disconnect( brokerkey ):
    global brokers
    if brokerkey=='': brokerkey='default'
    if brokerkey in brokers:
        brokers[brokerkey].stop()
        del brokers[brokerkey]

def disconnectall():
    global brokers
    for key,value in list(brokers.items()):
        disconnect( key )
        
## COMMANDS

"""
def exist( args ):
    if len(args)<1:
        die( COLOR(YELLOW)+"Missing parameter:", "exist <key>", 7 )
    key = args[0]
    default = False
    print( len(args))
    if len(args)==2:
        print( "has default flag" )
        temp = args[2][0]
        print(temp)
        default = True if args[2][0].lower()==true else False
    result = config.exist( key, default )
    print( COLOR(WHITE)+str(result) )
"""

def get( args ):
    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', cmd+' <key>', len(cmd)+2)
        return
    key = args.pop(0)
    if len(args)>0:
        unexpected = " ".join(args)
        die( COLOR(YELLOW)+'Unexpected argument:', cmd+' '+key+' '+unexpected, len(cmd)+3+len(key)  )
        return
    print( COLOR(WHITE)+key+" = '"+str(config.get( key ))+"'" )
        
def help( args ):
    print( COLOR(WHITE)+"""
  get <key>              Displays a configuration key
  set <key> <value>      Sets a configuration key
  subscribe <topic>      Subscribes to a topic (Shows messages)
  publish <topic> <msg>  Publishes a message on a topic
  version                Shows the current version
  
  Alias:
  show <key>    Same as "get <key>"
  listen <topic>  Same as "subscribe <topic>"
  """)

"""
def list( args ):
    print( COLOR(WHITE)+"LIST" )

    # READ CONFIG.INI FOR INSTALLED COMPONENTS
    # READ MODULES.DAT FOR AVAILABLE COMPONENTS
    # SHOW LIST
"""

def publish( args ):

    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'publish <topic> <Message>', 9 )
        return
        
    topic = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'publish '+topic+' <Message>', 10+len(topic) )
        return
       
    message  = " ".join(args)
        
    broker = connect()
    broker.publish( topic, message )
    print( COLOR(GREEN)+"- Published to {}.".format( topic ) )
 
def set( args ):
    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'set <key> <value>', 5)
        return
    key = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', 'set '+key+' <value>', 6+len(key))
        return
    value = " ".join(args)
    config.set( key, value )
    config.save()
    print( COLOR(WHITE)+key+" = '"+str(value)+"'" )

def subscribe( args ):
    cmd = args.pop(0)
    if len(args)<1:
        die( COLOR(YELLOW)+'Missing parameter:', cmd+' <topic>', len(cmd)+2)
        return
    topic = args.pop(0)
    if len(args)>0:
        unexpected = " ".join(args)
        die( COLOR(YELLOW)+'Unexpected argument:', cmd+' '+topic+' '+unexpected, len(cmd)+3+len(topic)  )
        return
        
    broker = connect()
    broker.subscribe( topic )
   
#def unsubscribe( args ):
#    pass

def unknown( args ):
    die( COLOR(YELLOW)+"- Unknown command: {}".format(args[0]) )
        
#def update( args ):
#    print( COLOR(WHITE)+"UPDATE")

def version( args ):
    print( COLOR(WHITE)+"Version: "+VERSION )

# DAEMON
def start_daemon( args )
    print( "- Starting daemon" )
    pass

# COMMAND SWITCHER
CMDswitcher = { 
    "--help":help,
    "get":get,
    "help":help,
    "listen":subscribe,
    "publish":publish,
    "set":set,
    "show":get,
    "subscribe":subscribe,
    "version":version,
    "--daemon":start_daemon
}
# "list":list,
# "update":update, 

if __name__=="__main__":

    # PROCESS ARGUMENTS
    if len(sys.argv) <= 1:
        interactive = True
        repl()
        sys.exit(0)

    func = CMDswitcher.get( sys.argv[1], unknown )
    func( sys.argv[1:] )
